cloudinit/sources/DataSourceScaleway.py

```diff
@@ -210,7 +210,10 @@ def _get_data(self):
         if self._fallback_interface is None:
             self._fallback_interface = net.find_fallback_nic()
         try:
-            with EphemeralDHCPv4(self._fallback_interface):
+            with EphemeralDHCPv4(
+                self._fallback_interface,
+                tmp_dir=self.distro.get_tmp_exec_path(),
```

Comment:
I think having each caller sort out tmp_dir is going to be error prone.  Even at the lowest level, `tempdir(needs_exe=True) ` now has a hidden dependency on `dir=self.distro.get_tmp_exec_path()`.

I've got two thoughts:
- figure out how to tell temp_utils where the executable tmp dir is without a distro reference on init?
- add a distro.executable_tmpdir() which handles the needs_exe=True and executable dir requirements, dropping those aspects from tmpdir().   this adds a requirement for net to become distro-aware, which I'm actually surprised isn't already the case, but suspect it'll happen eventually as bringing up ephemeral dhcp is likely to have distro-specific aspects (e.g. distros removing dhclient at some point).

Reply:
Thanks, @cjp256, for the detailed and interesting discussion.

> * figure out how to tell temp_utils where the executable tmp dir is without a distro reference on init?

In the current state of the PR (commit 29833a75611f67d0f86b954bb78f53c17f14eba5), there is no reference to `distro` within `temp_utils`.

> [0]: add a distro.executable_tmpdir() which handles the needs_exe=True and executable dir requirements, dropping those aspects from tmpdir().

There is a new Distro method [get_tmp_exec_path](https://github.com/canonical/cloud-init/pull/1690/files#diff-2902e7f3123dd1e701482d6c60d6ca2e51d2dcbe3f643cb30be70727fde3fbfbR953) which handles the executable dir requirements. It checks if temp_dir with need_exe target tmp dir is in a noexec mount point (via a new function [temp_utils.get_tmp_ancestor](https://github.com/canonical/cloud-init/pull/1690/files#diff-6179e8b68a1cb97cb60b99f1c269a896292fc6dec97cb450cec13886a0101e80R18)) and if so fallback to `{distro.usr_lib_exec}/cloud-init/clouddir`.

We could move the `need_exe` from `temp_utils` to `distro.get_tmp_exec_path` simplifying things a bit.

> [1]: this adds a requirement for net to become distro-aware, which I'm actually surprised isn't already the case, but suspect it'll happen eventually as bringing up ephemeral dhcp is likely to have distro-specific aspects (e.g. distros removing dhclient at some point).

I do not see how [0] implies [1], but assuming it does, I had some arguments against tightly coupling `distro` with `net.dchp` and/or `net.ephemeral`:

1. The necessity of copying and running `dchp` from a tmp dir in order to avoid apparmor restrictions is going to be eliminated (@holmanb can point to links about this, I believe). This implies the dependency will disappear.
2. After [WIP-ONGOING-REFACTORIZATION.rst](https://github.com/canonical/cloud-init/blob/main/WIP-ONGOING-REFACTORIZATION.rst), `net.dchp` and `net.ephemeral` are going to belong to a `cloudinit.distros.networking` hierarchy of classes mapped to the distro hierarchy. If the number of distro-specific aspects (to bring up ephemeral dchp) increases, we could encode the dependencies via this hierarchy.
3. I remember a discussion about dchp. I think `dchp` is going to be discontinued and we talk about the possibility of including a new dchp python-dependency and other alternatives. This will also remove the dependency. (But this belongs to the more obscure future)

Summarizing, I do not think that to only have a distro's exec tmp dir and net justifies to tightly couple them, but I am open to hear more arguments. Maybe @blackboxsw, @holmanb and @TheRealFalcon have opinions/arguments about this.

Reply:
@aciba90 @cjp256 I'm glad we're thinking about architecture, refactoring, and future direction here. Good points made all around.

The topic of dhcp (both sandboxing issues and the looming issue of dhclient's support status) has come up on irc periodically and maybe also in PR conversations. This topic probably deserves better community coordination and communication, since the problem has implications for clouds and distros (i.e. if we wanted to use a dedicated or auto-assigned link local ipv4 address, dhcp probably wouldn't be necessary, but does that work for all clouds?). This probably deserves a mailing list discussion. Or maybe we should find a way to get contributors in the same room to talk about this kind of stuff :)

> in order to avoid apparmor restrictions is going to be eliminated

This would eliminate some headaches and one historical source of bugs. However we might have to push on this one ourselves.

> Even at the lowest level, tempdir(needs_exe=True) now has a hidden dependency on dir=self.distro.get_tmp_exec_path().

Agreed, we already have implicit coupling.
It seems like more and more code has been shifting into the `distro` implementations recently. I expect this trend will likely continue, which suggests that `distro` may become required by more modules over time.

> 1. The necessity of copying and running dchp from a tmp dir in order to avoid apparmor restrictions is going to be eliminated (@holmanb can point to links about this, I believe). This implies the dependency will disappear.
> ...
> 3. I remember a discussion about dchp. I think dchp is going to be discontinued and we talk about the possibility of including a new dchp python-dependency and other alternatives. This will also remove the dependency. (But this belongs to the more obscure future)

Either of these solutions could potentially remove the requirement for a writable executable tempdir for networking. It does not, however, remove all dependency in cloud-init on `needs_exe`. I see at least a couple of other calls (though the debconf one does look unnecessary at a glance).

All in all, I'm not opposed to this approach. I do think that since we currently need distro input in this patch, it might be worth just creating a distro wrapper function to make the callsites less error prone and more DRY. Whether to do that before this PR lands or as a refactor after, I don't feel strongly about.

That's a lot of words to say basically "I don't have strong feelings on this. I could still be swayed either way." I'll wait in case @cjp256 @blackboxsw @TheRealFalcon or anyone else has more thoughts.

Reply:
To clarify, I'm not demanding any changes - I'm just driving by and sharing my thoughts fwiw. :D  I'm happy to have this fixed in whatever way you think is best at this time!

Reply:
@cjp256 Thanks much for sharing. Sounds good :)